From f260b0cfab0ac527be3b5f83b5d131d9abecfe2f Mon Sep 17 00:00:00 2001
From: Lyude Paul <lyude@redhat.com>
Date: Mon, 4 Mar 2024 19:01:10 -0500
Subject: [PATCH 021/109] rust: drm: Add a small handful of fourcc bindings

This adds some very basic rust bindings for fourcc. We only have a single
format code added for the moment, but this is enough to get a driver
registered.

Signed-off-by: Lyude Paul <lyude@redhat.com>

---
V3:
* Drop FormatList and ModifierList
  These aren't actually needed as pointed out by Louis Chauvet
* Add a constant for FORMAT_MOD_INVALID
  I realized that we actually need this because the format list isn't
  terminated with a 0 like I thought, and we can't pick this up
  automatically through bindgen
* Split out the FormatInfo WIP
  We'll want this someday, but not yet.

(cherry picked from commit d6cf66733c67c0679ccf81f3d139f9883608d4cc)
---
 rust/kernel/drm/fourcc.rs | 21 +++++++++++++++++++++
 rust/kernel/drm/mod.rs    |  1 +
 2 files changed, 22 insertions(+)
 create mode 100644 rust/kernel/drm/fourcc.rs

diff --git a/rust/kernel/drm/fourcc.rs b/rust/kernel/drm/fourcc.rs
new file mode 100644
index 000000000000..62203478b595
--- /dev/null
+++ b/rust/kernel/drm/fourcc.rs
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0 OR MIT
+
+//! DRM fourcc bindings.
+//!
+//! C header: [`include/uapi/drm/drm_fourcc.h`](srctree/include/uapi/drm/drm_fourcc.h)
+
+/// Return a fourcc format code.
+const fn fourcc_code(a: u8, b: u8, c: u8, d: u8) -> u32 {
+    (a as u32) | (b as u32) << 8 | (c as u32) << 16 | (d as u32) << 24
+}
+
+// TODO: We manually import this because we don't have a reasonable way of getting constants from
+// function-like macros in bindgen yet.
+#[allow(dead_code)]
+pub(crate) const FORMAT_MOD_INVALID: u64 = 0xffffffffffffff;
+
+// TODO: We need to automate importing all of these. For the time being, just add the single one
+// that we need
+
+/// 32 bpp RGB
+pub const XRGB888: u32 = fourcc_code(b'X', b'R', b'2', b'4');
diff --git a/rust/kernel/drm/mod.rs b/rust/kernel/drm/mod.rs
index fd6ed35bc35a..c58a3173830c 100644
--- a/rust/kernel/drm/mod.rs
+++ b/rust/kernel/drm/mod.rs
@@ -5,6 +5,7 @@
 pub mod device;
 pub mod driver;
 pub mod file;
+pub mod fourcc;
 pub mod gem;
 pub mod gpuvm;
 pub mod ioctl;
